kernel/pipedev.c

```diff
@@ -290,23 +285,33 @@ static ssize_t _pd_write(
     _lock(pipe);
     p = pipe->impl;
 
-    p->wrsize += count;
-
     if (!_valid_pipe(pipe))
     {
         /* cannot unlock since mutex is no longer valid */
         _unlock(pipe);
         ERAISE(-EBADF);
     }
 
+    /* The writer intends to write count bytes. Update wrsize immediately to
+     * signal to the reader about the intention, so a blocking reader won't bail
+     * out prematurely due 0 wrsize, when a writer will write more to the pipe
+     * soon. If the call returns with less than count bytes writtern, the caller
+     * might invoke the call again. For example, a non-blocking write returning
+     * EAGAIN did not write any data to the pipe. The caller likely will call
+     * again indicating count bytes to write. In those cases, adjust wrsize
+     * before returning to make sure no double counting
+     * */
+    p->wrsize += count;
```

Comment:
If we are here, it means the writer holds the lock, and there won't be readers checking `wrsize` at the same time. Why don't we increment `wrsize` when we actually know how many bytes are written to the pipe buffer?

Reply:
The writer will release a lock when the pipe is full and wait for other threads to read the pipe (line 345 in the new code).